feat[Core]: Admin and User Metadata for transactional ressources and users (#5897)

JustShah opened the pull request:
Closes #5897
Concerns #5951

# Breaking Changes
This feature breaks sqlite versions predating to [December 2023](https://sqlite.org/jsonb.html), verify your requirements before migrating. 

# Introduction 

## Overview
This PR introduces the concept of metadata fields to solidus. The purpose of metadata fields is to allow customers and admins alike to provide additional information to transactions to either extend the informations contained or allow to insert additional data to correlate transactions with third party systems. 

Examples for data can be: 

- any transaction itself such as orders, RMAs, payments and stock movements;
- individual items contained inside an order (line items);
- match users with 3rd party systems.

Possible Applications on a very high level are:

- customers that want to add Purchase Order information to an order; 
- administrators that want to add a ticket number to an RMA.

The current scope of meta data are following ressources:  
- Users
- Orders
- Customer Returns and any from of correlated transactions
- Payments
- Shipments
- Line Items

## How is the additional information stored?
for each of the above resources two columns have been created:
- customer_metadata
- admin_metadata

The resources store the information as jsonb.

## What is the access model?
Cancancan has been used to allow users to write data with the current scope just extended by the additional fields:
- Users can create / delete / alter metadata on orders only till the payment itself is processed,
- Admins have universal access R/W access to metadata at all stages of transactions.

## Testing
Extensive test cases have been added. 

# Detailed Overview 

## Types of Meta Data and usage suggestions

### Overview 
Meta data can be broken down in three large topics visible in the table below:

|                         | Use Case                                                                                                                                                              |
|-------------------------|-----------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| Administrative Metadata | Information required to correlate all transactions to 3rd party systems.                                                                                              |
| Customer Metadata       | Information allowing customers to customize orders (eg options to store information like a text string or link to an image) or correlate transactions (such as POs).  |

### Administrative Metadata
Transactional resources are either by nature integrated with 3rd party systems (eg Payments, Billing in Italy with eInvoicing) or by operational requirement (ERP, CRMs like Salesforce,...). Meta data as an extension on admin site for all transactional resources to connect the transactional resources to any of such systems. Given that more and more countries will switch to a form of eInvoicing in the coming years meta data allows to fetch or store additional information in those transactions.

### Customer Meta Data 
Customer meta data allows users to customize a cart with either administrative (PO numbers,...) or customisation information (eg an URL to a picture to customize a shirt, text strings for incisions or in our case IMEIs to relate to a subscription) allowing to significantly extend the features without changing the underlying model. Fields should be accessible on front-end and backend. 

## Schema 
Meta Data should net a reliable outcome: 
"key1" : "value1" should be as applicable as "key2" : "value2", without restriction other than the quantity of storable key value pairs and limitations of the dimension of the value given that json as a format does not pose limits we should do that server side. 

Future models to reserve or block certain key names might be possible. 

## Scopes & Arrays

### Nature of the Values
The system is agnostic to the value contained. A decision has been made to store the data in additional columns in a jsonb which by now is relatively widely supported. Some preexisting installations might have to upgrade DBs, older sqlite versions dating before Jan 2024 are therefor not supported and need to be updated.

"key1" : "value1" should lead to the same behaviour as "key2" : "value2" without requiring modification inside the code. Reasonable provisions should be taken inside the code base to edit the quantities and content size of key value pairs and appropriate errors should be returned where rate or size limits are exceeded. 

## Administrative Meta Data
Allow appropriate roles to customize resources with private fields not end-user accessible. 
R/W = Read and Write Access
|               | Admin   | Use Case                                                                                                                                  |
|---------------|---------|-------------------------------------------------------------------------------------------------------------------------------------------|
| **Order**     | **R/W** | **Add 3rd party system reference fields such as PO-Order or Dropshipping reference**                                                      |
| **Line Item** | **R/W** | **Add information to customize object such as link to an image for customization of a product or relate to a 3rd party system reference** |
| **RMA**       | **R/W** | **Add 3rd party system reference fields such as PO-Order or Dropshipping reference**                                                      |
| **Payment**   | **R/W** | **Add 3rd party system reference fields such as PO-Order or Dropshipping reference**                                                      |
| **Order**     | **R/W** | **Add 3rd party system reference fields such as PO-Order or Dropshipping reference**                                                      |
| Shipment      | R/W     | Add 3rd party system reference fields such as a delivery note                                                                             |
| Payment       | R/W     | Add 3rd party system reference fields for payment conciliation                                                                            |
| User       | R/W     | Add a 3rd party reference |

## Customer Meta Data 

Allowing users to customize an incomplete order (or I'd say a cart) and call the information after a completed order. Admins should in an ideal case be allowed to edit those information after purchase, but I feel that can wait if technical limitations are currently in place. 

|               | User    | Admin   |    Use Case                                                                                                                                       |
|---------------|---------|---------|-------------------------------------------------------------------------------------------------------------------------------------------|
| **Order**     | **R/W** | **R/W** | **Add 3rd party system reference fields such as PO-Order,**                                                                               |
| **Line Item** | **R/W** | **R/W** | **Add information to customize object such as link to an image for customization of a product or relate to a 3rd party system reference** |
| User          | R/W     | R/W     | Add additional User Fields or References                                                                                                  |

fthobe commented:
@shahmayur001 

Thomas' comment on slack: 
tvdeyen
  28 minutes ago
Just tried to review, but the changes in single commits are soo huge that they literaly crashed my browser.
Could you please create smaller commits that are self contained (ie. first introduce the actual metadata object, with tests but without using it and then commit by commit introduce it in the models that require it).
That way it would be much easier to review and reason about. There is a lot of good information in the PR description, but this information is missing in the commits. Writing smaller atomic commits helps also writing better commit messages, because the context is smaller and better explained that way.
This is a great resource I share often. Especially #7 is very important. We also linked some more notes about good git commit messages and why they matter in our CONTRIBUTION.md file.
https://cbea.ms/git-commit/

JustShah committed:
Add concern and validations for metadata with config options for keys and length

Introduced validations and configurable limits for metadata attributes
(`customer_metadata` and `admin_metadata`) to ensure consistency and prevent
overly large or complex structures.

**Implementation:**
- Set default values for `customer_metadata` and `admin_metadata` to empty hashes (`{}`) for proper initialization.
- Implemented validation to limit the number of keys, and the length of keys and values in metadata.
- Added configuration checks to ensure metadata does not exceed the limits defined in `Spree::Config`.
- Introduced preferences: `meta_data_max_keys`, `meta_data_max_key_length`, and `meta_data_max_value_length` to control metadata size:
  - `meta_data_max_keys`: Max keys allowed (default: 6).
  - `meta_data_max_key_length`: Max key length (default: 16).
  - `meta_data_max_value_length`: Max value length (default: 256).
- Created a shared concern to handle metadata validation logic.
- Added tests to ensure default values and validation logic work correctly.
- Created a shared example that can be used in other specs to ensure consistent testing of metadata validation across various parts of the system.

JustShah committed:
Include concern to models

Call the metadata concern to various models to validate and add admin and customer metadata

JustShah committed:
Add restriction for metadata update after order is complete

User cannot update customer_metadata after the status of order is complete

JustShah commented:
## permitted\_<resource\>_attributes Method (from base_controller.rb)
### Purpose:
This method is dynamically defined for each resource in the `METADATA_MODELS` array (e.g., `Order`, `CustomerReturn`, `Payment`, etc.).



### How it works:
- It checks if the current user has `:admin` permissions for the given resource (e.g., `Spree::Order`, `Spree::Payment`, etc.).
- If the user is an admin for that resource, it then appends `admin_metadata_attributes`.
- If the user is not an admin, it simply calls `super()` to return the default attributes for the resource.



### Key Use:
It ensures that admin users have additional metadata attributes, while non-admin users only get the default attributes.



---



## \<method\_name\>\_attributes Method (from api_helpers.rb)



### Purpose:
This method dynamically creates accessor methods for each resource in the `METADATA_MODELS` hash (e.g., `order`, `payment`, `user`, etc.).



### How it works:
- For each resource in the hash (e.g., `order: Spree::Order`, `payment: Spree::Payment`), it defines a method like `order_attributes`, `payment_attributes`, etc.
- Each generated method calls the `authorized_attributes` method, which checks if the user has `:admin` permissions for the resource.
- If the user is an admin, it adds `:admin_metadata` to the list of attributes defined in the `Spree::Api::Config` for that resource.
- If the user is not an admin, it returns only the default attributes for that resource.



### Key Use:
This method centralizes the logic for checking permissions and ensures that the appropriate set of attributes is returned based on user roles.



---



## Summary:



- The `permitted_<resource>_attributes` method from the `BaseController` dynamically defines which attributes are allowed for each resource, adding admin-specific metadata when appropriate.
- The `<method_name>_attributes` method in the helper module similarly checks user permissions and adjusts the set of attributes returned for API resources, appending admin metadata when the user has admin rights.

tvdeyen commented on core/lib/spree/app_configuration.rb:
```diff
@@ -209,6 +209,18 @@ class AppConfiguration < Preferences::Configuration
 
+    # @!attribute [rw] max_keys
+    #   @return [Integer] Maximum keys that can be allocated in customer and admin metadata column(default: +6+)
+    preference :max_keys, :integer, default: 6
```
Can you explain why this preference could be helpful?

fthobe replied:
It limits the maximum amount of key value pairs. 

As much xml expansion is not a problem here, we never the less given the low degree of auth needed to write practically anything create a stability issue. 

Limiting it seemed reasonable. I will see to adapt the PR and Documentation accordingly as i believe it's important to have it there.

kennyadsl replied:
I'm also not convinced this is a valid addition. Can you name one real world example where this might be needed?

tvdeyen commented on api/app/controllers/spree/api/orders_controller.rb:
```diff
@@ -112,12 +112,19 @@ def mine
+
+        params[:order].delete(:customer_metadata) if params[:order]
+      end
+
```
please add specs for this and explain why this is necessary in the commit message

JustShah replied:
We already have specs for this https://github.com/gms-electronics/solidus/pull/4/files#diff-2ef02110b947167ef3c10ad87001d8c36b1425578b4126fd5d39dd234e4906b7R132

Basically, This prevents users from updating customer_metadata once order it is completed

tvdeyen commented on api/app/controllers/spree/api/base_controller.rb:
```diff
@@ -5,6 +5,17 @@
+        :StockLocation,
+        :StockMovement,
+        :Variant
+      ]
```
This should be configurable via the app configuration.

tvdeyen replied:
please amend the change to the introducing commit

JustShah replied:
Resolved!

tvdeyen commented on api/app/helpers/spree/api/api_helpers.rb:
```diff
@@ -37,12 +37,33 @@ module ApiHelpers
+        stock_location: Spree::StockLocation,
+        stock_movement: Spree::StockMovement,
+        user: Spree::LegacyUser
+      }
```
This is duplicating the constant in the api controller. Can we make this configurable?

fthobe replied:
@shahmayur001 I'd love to see this, but estimate if we can slip this in.

JustShah replied:
Resolved!

tvdeyen commented on api/app/helpers/spree/api/api_helpers.rb:
```diff
@@ -37,12 +37,33 @@ module ApiHelpers
+        shipment: Spree::Shipment,
+        stock_location: Spree::StockLocation,
+        stock_movement: Spree::StockMovement,
+        user: Spree::LegacyUser
```
Please use `Spree.user_class` `Spree::LegacyUser` is only used in the dummy app for testing.

fthobe replied:
@shahmayur001 this one makes sense

JustShah replied:
Resolved!

tvdeyen commented on api/app/helpers/spree/api/api_helpers.rb:
```diff
@@ -37,12 +37,33 @@ module ApiHelpers
+
+      def authorized_attributes(resource, config_attribute)
+        can?(:admin, resource) ? Spree::Api::Config.public_send(config_attribute) + [:admin_metadata] : Spree::Api::Config.public_send(config_attribute)
+      end
```
This seems to duplicate what already is done in the api base controller (this module is mixed into). Why was this change necessary?

fthobe replied:
@shahmayur001 this one makes sense

JustShah replied:
We are adding the attribute to the params in API helpers, meanwhile, in the base controller, we are permitting the attribute.

kennyadsl replied:
I also do not understand why we need this. Why do we need this here if we already set `permitted__<resource\>_attributes`?

tvdeyen commented on api/spec/requests/spree/api/line_items_spec.rb:
```diff
@@ -24,7 +24,7 @@ module Spree::Api
     it "can learn how to create a new line item" do
       get spree.new_api_order_line_item_path(order)
-      expect(json_response["attributes"]).to eq(["quantity", "price", "variant_id"])
+      expect(json_response["attributes"]).to eq(["quantity", "price", "variant_id", "customer_metadata"])
```
the specs should be squashed into the same commit the change has been introduced in the controller

JustShah replied:
Resolved!

tvdeyen commented on api/spec/requests/spree/api/orders_spec.rb:
```diff
@@ -79,6 +79,68 @@ module Spree::Api
         end
       end
 
+      context "when the user is not admin but has ability to create and update orders" do
```
these specs also belong into the same commit as the implementation

JustShah replied:
Resolved!

tvdeyen commented on api/spec/requests/spree/api/payments_spec.rb:
```diff
@@ -45,6 +50,22 @@ module Spree::Api
             expect(json_response).to have_attributes(attributes)
           end
 
+          it "allows creating payment with customer metadata but not admin metadata" do
```
these specs also belong into the same commit as the implementation

JustShah replied:
Resolved!

tvdeyen commented on api/spec/requests/spree/api/users_spec.rb:
```diff
@@ -15,6 +15,20 @@ module Spree::Api
+
+        expect(json_response['email']).to eq user.email
+        expect(json_response).not_to have_key('admin_metadata')
+      end
```
these specs also belong into the same commit as the implementation

JustShah replied:
Resolved!

tvdeyen commented on core/app/models/concerns/spree/metadata.rb:
```diff
@@ -0,0 +1,52 @@
+    extend ActiveSupport::Concern
+
+    included do
+      store :customer_metadata, coder: JSON
```
We need to tell active record to always use `json` as column type in order to work in maria db

See https://github.com/AlchemyCMS/alchemy_cms/commit/0817f1d7742adfc018c750a221b89e73303fd76d

```rb
attribute :customer_metadata, :json
```

fthobe replied:
@shahmayur001 makes sense, please adjust

JustShah replied:
Resolved!

tvdeyen commented on core/db/migrate/20250129061658_add_metadata_to_spree_resources.rb:
```diff
@@ -22,8 +22,8 @@ def change
-          add_column table_name, :customer_metadata, :json, default: {}
-          add_column table_name, :admin_metadata, :json, default: {}
+          add_column table_name, :customer_metadata, :json
+          add_column table_name, :admin_metadata, :json
```
This should be amended into the first commit

fthobe replied:
@shahmayur001 sqaushing commits that are redundant makes sense, please adjust

JustShah replied:
Resolved!

tvdeyen reviewed (commented):
this pr is still very hard to review, because the commits lack the information on why a certain change has been necessary. also lots of changes seem to be fixups for previous commits. this is hard to follow sometimes. I would also like to keep the list of models we introduce this into to a bare minimum. we can either add them later or (better) create a generator that allows to easily integrate it in other models if necessary.

fthobe commented:
@tvdeyen can we agree on all transactional models plus users?

- Orders
- Shipments
- Payments
- Users
- Line Items
- RMAs and Returns

I think that would be good enough for now and we can later decide to change addresses as well.

tvdeyen commented:
> - Orders
> - Shipments
> - Payments
> - Users
> - Line Items
>
> I think that would be good enough for now and we can later decide to change addresses as well.

Yes

fthobe commented:
@shahmayur001 to not throw the work, please squash commits that overwrite each other, keep an eye on the comments that @tvdeyen has made related to DB compatibility, we really missed that one and that was an excellent catch. Remove everything that is not transactional but users and in addition we remove stock movements as people can work (and probably will) push stock only from ERP and keep delivery notes outside.

@tvdeyen I would keep returns / rmas for the same reasons as I think that this is a needed feature at the first place (correlation with 3rd party systems). Stock movements I believe are overkill. Can you live with that?

tvdeyen commented:
Yes, we can live with that. Stock movements table can get pretty big, pretty fast. That's why we want to keep it as simple as possible.

fthobe commented:
@tvdeyen I can see that. 

@shahmayur001 so the agreed upon version is to also keep RMA / returns.

fthobe commented:
Hey @tvdeyen 

Just to get the vibe here, as this is not something that is reasonably easy broken down in atomic commits. In my opinion it should be broken down as follows: 
- everything concerning model and DB migrations
- API 
- Testing 
- Lint 

Are 4 commits ok here?

fthobe commented on core/db/migrate/20250129061658_add_metadata_to_spree_resources.rb:
```diff
@@ -0,0 +1,32 @@
+      spree_customer_returns
+      spree_stock_locations
+      spree_store_credit_events
+      spree_variants
```
We already decided to remove spree_variants from the meta data

tvdeyen replied:
Actually, on second thought. Variants could be a good candidate for structured meta data.

fthobe replied:
@tvdeyen can you brake out why and we take a look at it? 

I assume that you want to show content about variants on frontend which was a discussed feature (public catalogue meta data). I would put that in a different PR. 

Metadata with public view access for storefront that is admin edititable. But if you have a different use case I am really curious about it.

tvdeyen replied:
I think of additional warehousing related (row, bay, bin,m packaging) and physical attributes for ie. calculations on ingredients (candle recipes and so on) that are not necessarily customer facing.

tvdeyen replied:
But we can introduce this in a separate PR. Due to the fact that this is a mixin we can easily split up the work.

fthobe replied:
> I think of additional warehousing related (row, bay, bin,m packaging) and physical attributes for ie. calculations on ingredients (candle recipes and so on) that are not necessarily customer facing.

I can see that. Let's push the doctrin through and than we can focus on refining.

fthobe replied:
> I think of additional warehousing related (row, bay, bin,m packaging) and physical attributes for ie. calculations on ingredients (candle recipes and so on) that are not necessarily customer facing.

It's actually interesting, I haven't thought about picking on this, but it makes totally sense.

tvdeyen commented:
> Hey @tvdeyen
> 
> Just to get the vibe here, as this is not something that is reasonably easy broken down in atomic commits. In my opinion it should be broken down as follows:
> 
> * everything concerning model and DB migrations
> * API
> * Testing
> * Lint
> 
> Are 4 commits ok here?

This is how I imagine how this could be done

1. Introduce the new mixin/concern first
2. Then it is easy to introduce the mixin into one or several models, together with migrations and tests of course (thats important for git bisects, etc.)
2. API introduction can be (a) separate commit(s), but also with tests
3. Linting of newly introduced files should be done during introduction of the file (ie. with an auto-formater). I recommend using the ruby-lsp extension for VSCode and its rubocop auto formatting feature.
4. Linting changes of foreign code can be done in a separate commit

All of these are just how it could be done, not hard rules. There are several flavours and styles of using git for coding. I like to think of using git as if you were telling a story. Introduction, interaction, changes, etc. The most important part is to write the why a change has been introduced in the commit message.

There are several great examples of well written PRs from the folks of Super Good and Martin in the Solidus repo.

I also shared a lot of examples via Slack and google/duckduckgo are also your friends.

fthobe commented on core/db/migrate/20250129061658_add_metadata_to_spree_resources.rb:
```diff
@@ -0,0 +1,31 @@
+class AddMetadataToSpreeResources < ActiveRecord::Migration[7.2]
+  def change
+    # List of Resources to add metadata columns to
+    %i[
```
Please apply metadata only to
      spree_users
      spree_orders
      spree_line_items
      spree_shipments
      spree_payments
      spree_refunds
      spree_customer_returns
      spree_store_credit_events
      spree_return_authorizations

fthobe commented:
@shahmayur001 I think this looks really really good. Consolidate the PR description and we open a PR after approval by Thomas

fthobe commented:
> > Hey @tvdeyen
> > Just to get the vibe here, as this is not something that is reasonably easy broken down in atomic commits. In my opinion it should be broken down as follows:
> > 
> > * everything concerning model and DB migrations
> > * API
> > * Testing
> > * Lint
> > 
> > Are 4 commits ok here?
> 
> This is how I imagine how this could be done
> 
> 1. Introduce the new mixin/concern first
> 2. Then it is easy to introduce the mixin into one or several models, together with migrations and tests of course (thats important for git bisects, etc.)
> 3. API introduction can be (a) separate commit(s), but also with tests
> 4. Linting of newly introduced files should be done during introduction of the file (ie. with an auto-formater). I recommend using the ruby-lsp extension for VSCode and its rubocop auto formatting feature.
> 5. Linting changes of foreign code can be done in a separate commit
> 
> All of these are just how it could be done, not hard rules. There are several flavours and styles of using git for coding. I like to think of using git as if you were telling a story. Introduction, interaction, changes, etc. The most important part is to write the why a change has been introduced in the commit message.
> 
> There are several great examples of well written PRs from the folks of Super Good and Martin in the Solidus repo.
> 
> I also shared a lot of examples via Slack and google/duckduckgo are also your friends.

We are close enough I believe and can get closer on future commits. 
https://github.com/gms-electronics/solidus/pull/4/commits

kennyadsl reviewed (commented):
Agree with all of what Thomas mentioned.

JustShah committed:
Introduce Customer and Admin Metadata Restriction Preference

This configuration adds the ability to enable or disable metadata restrictions as needed.

Following configurations are possible:
- `meta_data_validation_enabled` : Enables or disables restrictions globally.

JustShah committed:
Linting

JustShah committed:
Add metadata columns for resources

This migration adds customer_metadata and admin_metadata to various tables

JustShah committed:
Add admin_metadata to attributes for admins

Introduce the `admin_metadata` attribute to support dynamic metadata handling for admins.
It enables configuration of metadata via API preferences, and ensures that metadata fields
are properly permitted for admin users through new helper methods.

**Implementation:**
- Added `admin_metadata` to resources for admins.
- Integrated `metadata_permit_parameters` and `metadata_api_parameters preference` in API configuration.
- Implemented `permitted_<resource>_attributes` and `<method_name>_attributes` methods for dynamic metadata permissions.
- Added `extract_metadata` in `line_items_controller` to handle metadata fields.

JustShah committed:
Permit `customer_metadata` attributes for APIs and add support for metadata handling

Update the API configuration to permit the `customer_metadata` attribute across various models,
such as orders, payments, shipments, and return authorizations. This change allows better handling of customer metadata within API requests,
ensuring consistent storage and validation of metadata attributes. Additionally, it updates the `preference` configurations for models to include `customer_metadata`.

**Implementation:**
- Added `customer_metadata` to the permitted attributes for several API resources.
- Updated various `preference` settings, including `order_attributes`, `payment_attributes`, `shipment_attributes`, etc., to include `customer_metadata`.
- Ensured that `customer_metadata` is correctly permitted in the API parameters for each relevant model, allowing admins and users to store custom metadata.
- This change improves the flexibility of metadata management within the API, enabling dynamic handling and easier customizations.
